AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/InjectHelper.java

```diff
@@ -0,0 +1,200 @@
+import javax.lang.model.element.VariableElement;
+
+import org.androidannotations.ElementValidation;
+import org.androidannotations.holder.GeneratedClassHolder;
+import org.androidannotations.holder.HasMethodInjection;
+import org.androidannotations.holder.HasMethodInjectionWrapper;
+
+import com.helger.jcodemodel.IJExpression;
+import com.helger.jcodemodel.IJStatement;
+import com.helger.jcodemodel.JBlock;
+import com.helger.jcodemodel.JExpr;
+import com.helger.jcodemodel.JFieldRef;
+import com.helger.jcodemodel.JInvocation;
+
+public class InjectHelper<T extends GeneratedClassHolder> {
+
+	public enum ValidationMode {
+		ENHANCED, VIEW_SUPPORT, BEAN, ACTIVITY_OR_FRAGMENT, ACTIVITY, FRAGMENT;
+	}
+
+	private final Map<ExecutableElement, List<ParamHelper>> methodParameterMap = new HashMap<>();
+
+	private final ValidatorHelper validatorHelper;
+	private final HasMethodInjection<T> handler;
+	private ValidationMode validationMode;
+
+	public InjectHelper(ValidatorHelper validatorHelper, HasMethodInjection<T> handler) {
+		this(validatorHelper, handler, ValidationMode.ENHANCED);
+	}
+
+	public InjectHelper(ValidatorHelper validatorHelper, HasMethodInjection<T> handler, ValidationMode validationMode) {
+		this.validatorHelper = validatorHelper;
+		this.validationMode = validationMode;
+		this.handler = handler;
+	}
+
+	public void validate(Class<? extends Annotation> expectedAnnotation, Element element, ElementValidation valid) {
+		Element enclosingElement = element.getEnclosingElement();
+		if (element instanceof VariableElement && enclosingElement instanceof ExecutableElement) {
+			validatorHelper.param.annotatedWith(expectedAnnotation).multiple().validate((ExecutableElement) enclosingElement, valid);
+			validatorHelper.doesNotHaveAnyOfSupportedAnnotations(enclosingElement, valid);
+			validateEnclosingElement(enclosingElement, valid);
+
+		} else if (element instanceof ExecutableElement) {
+			validateEnclosingElement(element, valid);
+			validatorHelper.param.anyType().validate((ExecutableElement) element, valid);
+			List<? extends VariableElement> parameters = ((ExecutableElement) element).getParameters();
+			for (VariableElement param : parameters) {
+				validatorHelper.doesNotHaveAnyOfSupportedAnnotations(param, valid);
+			}
+
+		} else {
+			validateEnclosingElement(element, valid);
+		}
+	}
+
+	public Element getParam(Element element) {
+		if (element instanceof ExecutableElement) {
+			return ((ExecutableElement) element).getParameters().get(0);
+		}
+		return element;
+	}
+
+	public void process(Element element, T holder) {
+		if (element instanceof ExecutableElement) {
+			processMethod(element, holder);
+		} else {
+			Element enclosingElement = element.getEnclosingElement();
+			if (enclosingElement instanceof ExecutableElement) {
+				processParam(element, holder);
+			} else {
+				processField(element, holder);
+			}
+		}
+	}
+
+	private void validateEnclosingElement(Element element, ElementValidation valid) {
```

Comment:
moved that method into `MethodInjectionHandler` and implemented the validation in each handler